### PR TITLE
[IA-4313] Delete App records if workspace no longer exists

### PR DESCRIPTION
**Pull request by lucymcnatt**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/[ticket_number]

### What

When an MRG is deleted in Azure, the workspace is deleted as well, so when Leo asks for the workspaceDescription from WSM, a 404 is returned. Instead of erroring, Leo's app records should be cleaned up in this case because if the workspace has been deleted, none of the resources under it exist anymore.

### Why

A workspace should be able to be properly cleaned up in Terra, even if the MRG is removed from under it

### What to test

- 

### Who tested and where

- [ ] This change is covered by automated tests 
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [ ] I validated this change 
- [ ] Primary reviewer validated this change 
- [ ] I validated this change in the dev environment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppService.scala [hunk 1]
```diff
@@ -37,7 +37,7 @@ trait AppService[F[_]] {
     as: Ask[F, AppContext]
   ): F[Vector[DiskName]]
 
-  def deleteAppRecords(userInfo: UserInfo, cloudContext: CloudContext.Gcp, appName: AppName)(implicit
+  def deleteAppRecords(userInfo: UserInfo, cloudContext: CloudContext, appName: AppName)(implicit
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppService.scala:40]: Oh nice! Glad that can be reused!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala [hunk 2]
```diff
@@ -820,7 +822,22 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
...
+
+    workspaceApi <- wsmClientProvider.getWorkspaceApi(userInfo.accessToken.token)
+    attempt <- F.delay(workspaceApi.getWorkspace(workspaceId.value, IamRole.READER)).attempt
+
+    _ <- attempt match {
+      // if the workspace is found, delete the app normally
+      case Right(workspaceDesc) =>
+        deleteAppV2Base(appResult.app, appResult.cluster.cloudContext, userInfo, workspaceId, deleteDisk, workspaceDesc)
+      // if the workspace can't be found, delete the workspace records
+      case Left(_: NotFoundException) =>
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala]: One thing to double check though is that the `NotFoundException` does not also corresponds with a Permission Denied. 

I know that internally within Leo we often return a RuntimeNotFound when a user does not have permission for instance

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala [hunk 3]
```diff
@@ -831,23 +848,53 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
...
+
+    _ <- attempt match {
+      // if the workspace is found, delete the app normally
+      case Right(workspaceDesc) =>
+        apps
+          .traverse { app =>
+            deleteAppV2Base(app._1, app._2, userInfo, workspaceId, deleteDisk, workspaceDesc)
+          }
+      // if the workspace can't be found, delete the workspace records
+      case Left(_: NotFoundException) =>
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala]: Same comment as above re triple checking the NotFoundException only corresponds to a 404

File: project/Dependencies.scala [hunk 4]
```diff
@@ -150,13 +150,13 @@ object Dependencies {
   def excludeSpringBoot = ExclusionRule("org.springframework.boot")
   def excludeSpringAop = ExclusionRule("org.springframework.spring-aop")
   def excludeSpringData = ExclusionRule("org.springframework.data")
-  def excludeSpringFramework = ExclusionRule("org.springframework")
+  // def excludeSpringFramework = ExclusionRule("org.springframework")
```
Comment (lucymcnatt) [inline project/Dependencies.scala]: this change was necessary when I was writing the tests...apparently the org.springframework.HttpStatus was needed to throw a NotFoundException

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala (hunk 3 above)
Comment (lucymcnatt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala]: we do already check workspace permission on line 848 (and in the other just deleteAppV2 function)
`_ <- F.raiseUnless(hasWorkspacePermission)(ForbiddenError(userInfo.userEmail))`

but yes on the workspace swagger there's a separate `403 Permission denied` response

**Review by LizBaldo — APPROVED** [review]

**Pull request merged**